src/main/kotlin/com/autonomousapps/subplugin/ProjectPlugin.kt

```diff
@@ -143,7 +142,7 @@ internal class ProjectPlugin(private val project: Project) {
       }
 
       unitTestSourceSets?.let { sourceSets ->
-        val variantSourceSet = newVariantSourceSet(name, SourceSetKind.TEST, sourceSets)
+        val variantSourceSet = newVariantSourceSet(name, SourceSet.TEST_SOURCE_SET_NAME, sourceSets)
```

Comment:
For Android, we might (need to) keep the special handling of each source set we support (at leas initially?). Does adding arbitrary source sets make sense in Android anyway?

Reply:
Android can have arbitrary source sets up to a point. There are some docs [here](https://developer.android.com/studio/build/build-variants).

As far as I can tell, an Android project has support for source sets like the following:
```
// all standard. note that src/test and src/androidTest extend from src/main, 
// but do not relate to each other. This is an important point for computing 
// the advice in StandardTransform
src/main
src/test
src/androidTest

src/debug
src/release
src/flavorDebug
src/flavor1Flavor2Release
... etc.
```
In particular, note that only the production code can have variant-specific source-sets. Both `test` and `androidTest` do not have variants (I think). You can see some more examples [here](https://developer.android.com/studio/build/build-variants#sourceset-build).

Finally, I think it is highly unusual and possibly "impossible" for an Android project to have truly arbitrary source sets. The source sets you will find all relate precisely to some known combination of flavors and build types.

Reply:
Thanks for the explanation. Took me a bit to wrap my head around this again. So the Android Plugin has its own SourceSet implementation and you can't see them when you do `the<SourceSetContainer>()`. (Not confusing at all when discussing this topic...)

But conceptually, from the dependency analysis perspective, both _source set_ concepts are similar.

I think if we put enough thought into this, it could be unified how the plugin handles any of these source sets and the associated configurations.

Do I understand correctly that the additional **Android** source sets are also not handled correctly right now?

For example, if I put code into "src/debug", that is not considered during the analysis, right? Or is there some special handling for this already?

Reply:
Never mind. For Android there is one `Android*Analyzer` for each _source set_ already. 

Maybe we can then basically transfer that approach to the JVM world. 🤔 

What would need to be generalised somehow is the relationship between source sets. At the moment relationships between _source sets_ are hardcoded. For example between "Main" and "Test" (Test sees dependencies of Main) - for which `SourceSetType` is used right now IIUC.
I presume the relationship between "main" and "debug" (Debug seems dependencies of Main) is also encoded somewhere.

I am thinking that the analysis needs additional inputs like this:
- What does the "compileClasspath" extend from? From this we could find out which dependencies are included (e.g. Test includes "Main" and "Test")
- The list of all sourceSetNames (need to be computed differently for Android and JVM) from which we can make a better mapping from a configuration name to a "Variant" (and with that the associated compile tasks).
 
Then this can be fixed:

https://github.com/autonomousapps/dependency-analysis-android-gradle-plugin/blob/03a4ad7e6655beaecb75b1dabd85c4b87af08d8d/src/main/kotlin/com/autonomousapps/model/declaration/Configurations.kt#L40-L41

Reply:
An 'issue' I face in my project is actually that I sometimes get wrong advice (e.g. for testFixture dependencies). 

It might already be a good improvement to address the TODO above by ignoring the unsupported source sets for now. Something that is done for 'androidTest' explicitly in some other place in the code already I think.

If we pass the names of the supported source sets into the analysis, this can probably work. I will give it a try and I might open a PR only for that. Might be a good starting point.

Reply:
Some clarifications:
> I presume the relationship between "main" and "debug" (Debug seems dependencies of Main) is also encoded somewhere.

In fact, this plugin never analyzes an Android project's "main" source directly. It analyzes every available _variant_ of an Android project. For simplicity, imagine there are two (the most common case): `debug` and `release`. Therefore the plugin analyzes the debug variant and the release variant. The plugin's integration with AGP ensures it analyzes all the relevant source sets for each of these variants.

> What does the "compileClasspath" extend from? From this we could find out which dependencies are included (e.g. Test includes "Main" and "Test")

I'm a little worried this won't work as expected. A few months ago I tried building a DAG of configurations based on `extendsFrom` relationships, whereupon I discovered that AGP does not use that consistently. There are a number of "floating" configurations that have logical relationships with each other, that are _not_ encoded with any real API, but rather custom handling inside AGP. The relationships are deterministic, however, which is why I just hand-code some algorithms for managing all that.

> It might already be a good improvement to address the TODO above by ignoring the unsupported source sets for now. Something that is done for 'androidTest' explicitly in some other place in the code already I think.

I agree.

Reply:
> I tried building a DAG...

If you have already been down that road, then it is probably not helpful to pursue this further.
Conceptually, I also wonder if this "extendsFrom" approach is even something to recommend. It always was like this for normal "test" dependencies in Gradle (which it also got from Maven I presume).
in case of _custom source sets_, It is very unclear what has an effect on what if you start to add your own "extendsFrom" statements. Because you, as the person that defines dependencies, would also need to know the extendsFrom-hierarchy to understand what you are doing. Which is quite hidden, if it does no longer follow the standard api/implementation etc. setup.
If I define my own source sets (e.g. for integration tests), I always prefer to not do that. And instead declare the needed dependencies explicitly again. Which is much in line what the plugin proposes I think -> Always directly declare what you use.
Maybe for _custom source sets_ the plugin should not attempt to derive anything from addition _extendsFrom_. And just treat them as if they are 'isolated'. Maybe later the plugin could offer a configuration option were users can explicitly say that they do not want declare dependencies for a certain source set if they are declared in for another source set already. 

> I agree.

Hope this goes in the right direction: #651